k2/csrc/host_shim_inl.h

```diff
@@ -0,0 +1,131 @@
+/**
+ * @brief
+ * host_shim_inl
+ *
+ * @note
+ * Don't include this file directly; it is included by host_shim.h.
+ * It contains implementation code.
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Haowen Qiu)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#ifndef K2_CSRC_HOST_SHIM_INL_H_
+#define K2_CSRC_HOST_SHIM_INL_H_
+
+#ifndef IS_IN_K2_CSRC_HOST_SHIM_H_
+#error "this file is supposed to be included only by host_shim.h"
+#endif
+
+#include <algorithm>
+#include <limits>
+#include <utility>
+#include <vector>
+
+#include "k2/csrc/array.h"
+#include "k2/csrc/fsa.h"
+#include "k2/csrc/host/weights.h"
+#include "k2/csrc/host_shim.h"
+#include "k2/csrc/ragged.h"
+
+namespace k2 {
+namespace internal {
+// Will be used in below functions `GetForwardScores` and `GetBackwardScores` to
+// convert Array1<double> to Array1<float> (then return it) or return itself.
+// May delete this finally as we just use it for current test purpose.
+template <typename FloatType>
+Array1<FloatType> ConvertArrayType(const Array1<double> &src);
+template <>
+inline Array1<double> ConvertArrayType<double>(const Array1<double> &src) {
+  K2_CHECK_EQ(src.Context()->GetDeviceType(), kCpu);
+  return src;
+}
+template <>
```

Comment:
seems to me this could go in array_ops{,_inl}.h.

Reply:
is this really needed by other functions? I was thinking it's only used here so i put it here in a internal namespace.

Reply:
Actually if we were to implement this I think it would be better to have it a function Astype(), in class Array.
But not necessary right now; just for the future.

Reply:
Added AsType in Array1